src/py42/exceptions.py

```diff
@@ -63,6 +63,44 @@ def __init__(self, error_message):
         super(Py42StorageSessionInitializationError, self).__init__(error_message)
 
 
+class Py42UserAlreadyAddedError(Py42Error):
+    """An exception raised when the user is already added to group or list, such as the
+    Departing Employee list."""
+
+    def __init__(self, user_id, list_name):
+        msg = u"User with ID {} is already on the {}.".format(user_id, list_name)
+        super(Py42UserAlreadyAddedError, self).__init__(msg)
+
+
+class Py42LegalHoldNotFoundOrPermissionDeniedError(Py42Error):
+    """An exception raised when a legal hold matter is inaccessible from your account."""
+
+    def __init__(self, matter_id):
+        super(Py42LegalHoldNotFoundOrPermissionDeniedError, self).__init__(
+            u"Matter with id={} either does not exist or your account does not have permission to view it.".format(
+                matter_id
+            )
+        )
+
+
+class Py42UserDoesNotExistError(Py42Error):
+    """An exception raised when a username is not in our system."""
+
+    def __init__(self, username):
+        super(Py42UserDoesNotExistError, self).__init__(
+            u"User '{}' does not exist.".format(username)
+        )
+
+
+class Py42InvalidRuleTypeError(Py42Error):
+    """An exception raised when trying to add or remove users to a system rule."""
+
+    def __init__(self, rule_id, source):
+        msg = u"Only alert rules with a source of 'Alerting' can be targeted by this command. "
+        msg += u"Rule {0} has a source of '{1}'."
+        super(Py42InvalidRuleTypeError, self).__init__(msg.format(rule_id, source))
+
+
```

Comment:
Can we have these be subclasses of Py42HTTPError and keep the original exception's response data on them? 

Edit: changed the selection range, not sure why it picked that part initially.

Reply:
Couldn't just about every exception be a Py42HTTPError, even ones prior to this PR such as `Py42ChecksumNotFoundError` or `Py42SessionInitializationError` ?

Reply:
Also, `Py42UserDoesNotExist` does not actually arise from an exception. It happens in `get_by_username()` which returns an empty list when the user is not found...

I guess that is a change I should probably document.

Reply:
Can we wrap the empty list response in one? I think anytime we're raising an exception as a result of making an http request, we should have some way of inspecting the request/response object itself.

Reply:
Sorry, somehow missed your first comment. 

Yeah, from a debugability perspective, I would prefer any error that results from an http request to contain the request data with it and not eat it. Because if, for some reason, there's a bug and we're raising the exception when we shouldn't, then we would need to modify py42 first (or use other tools to make the same request and inspect it) to be able to figure out why.

Reply:
Makes sense to me... I got it working currently but am not too satisfied with how the exceptions module behaves, so will look into refactoring